File: src-tauri/src/discovery/project_scanner.rs
/// Project scanning — discover projects from ~/.claude/projects/.

use std::collections::HashMap;
use std::io::{BufRead, BufReader};
use std::path::Path;

use crate::types::domain::Project;
use crate::types::jsonl::RawJsonlEntry;

use super::content_filter::has_non_noise_messages;
use super::path_decoder::{decode_path, extract_project_name, is_valid_encoded_path};
use super::subproject_registry::SubprojectRegistry;

/// Scan the projects directory and return all projects.
pub fn scan_projects(
    projects_dir: &Path,
    registry: &mut SubprojectRegistry,
) -> Result<Vec<Project>, String> {
    if !projects_dir.exists() {
        return Ok(vec![]);
    }

    registry.clear();
    let mut all_projects = Vec::new();

    let entries = std::fs::read_dir(projects_dir)
        .map_err(|e| format!("Failed to read projects dir: {e}"))?;

    for entry in entries.flatten() {
        let name = entry.file_name();
        let name = name.to_string_lossy().to_string();

        if !entry.path().is_dir() || !is_valid_encoded_path(&name) {
            continue;
        }

        match scan_project(projects_dir, &name, registry) {
            Ok(projects) => all_projects.extend(projects),
            Err(e) => {
                eprintln!("[scanner] Error scanning project {name}: {e}");
            }
        }
    }

    // Sort by most recent session (descending)
    all_projects.sort_by(|a, b| {
        let a_time = a.most_recent_session.unwrap_or(0.0);
        let b_time = b.most_recent_session.unwrap_or(0.0);
        b_time.partial_cmp(&a_time).unwrap_or(std::cmp::Ordering::Equal)
    });

    Ok(all_projects)
}

/// Scan a single project directory. May return multiple projects if
/// sessions have different `cwd` values (subproject splitting).
fn scan_project(
    projects_dir: &Path,
    encoded_name: &str,
    registry: &mut SubprojectRegistry,
) -> Result<Vec<Project>, String> {
    let project_dir = projects_dir.join(encoded_name);
    let entries = std::fs::read_dir(&project_dir)
        .map_err(|e| format!("Failed to read {}: {e}", project_dir.display()))?;

    // Collect session files with their cwds
    let mut sessions_by_cwd: HashMap<String, Vec<(String, f64)>> = HashMap::new();
    let mut default_cwd: Option<String> = None;

    for entry in entries.flatten() {
        let file_name = entry.file_name();
        let file_name = file_name.to_string_lossy();

        if !file_name.ends_with(".jsonl") || file_name.starts_with("agent_") {
            continue;
        }

        let session_id = file_name.trim_end_matches(".jsonl").to_string();
        let file_path = entry.path();

        // Get file creation time for sorting
        let created_at = entry
            .metadata()
            .ok()
            .and_then(|m| m.created().ok())
            .and_then(|t| t.duration_since(std::time::UNIX_EPOCH).ok())
            .map(|d| d.as_secs_f64() * 1000.0)
            .unwrap_or(0.0);

        // Extract cwd from the first line of the session file
        let cwd = extract_cwd_from_file(&file_path).unwrap_or_default();

        if default_cwd.is_none() && !cwd.is_empty() {
            default_cwd = Some(cwd.clone());
        }

        let key = if cwd.is_empty() {
            "__default__".to_string()
        } else {
            cwd
        };

        sessions_by_cwd
            .entry(key)
            .or_default()
            .push((session_id, created_at));
    }

    // If all sessions have the same cwd (or only one group), return one project
    if sessions_by_cwd.len() <= 1 {
        let (cwd, sessions) = sessions_by_cwd
            .into_iter()
            .next()
            .unwrap_or(("__default__".to_string(), vec![]));

        let cwd_hint = if cwd == "__default__" {
            default_cwd.as_deref()
        } else {
            Some(cwd.as_str())
        };

        let session_ids: Vec<String> = sessions.iter().map(|(id, _)| id.clone()).collect();
        let most_recent = sessions.iter().map(|(_, t)| *t).fold(0.0f64, f64::max);
        let created_at = sessions.iter().map(|(_, t)| *t).fold(f64::MAX, f64::min);

        return Ok(vec![Project {
            id: encoded_name.to_string(),
            path: cwd_hint
                .map(|s| s.to_string())
                .unwrap_or_else(|| decode_path(encoded_name)),
            name: extract_project_name(encoded_name, cwd_hint),
            sessions: session_ids,
            created_at: if created_at == f64::MAX { 0.0 } else { created_at },
            most_recent_session: if most_recent > 0.0 {
                Some(most_recent)
            } else {
                None
            },
        }]);
    }

    // Multiple cwds → create composite projects
    let mut projects = Vec::new();

    for (cwd, sessions) in sessions_by_cwd {
        let cwd_str = if cwd == "__default__" {
            default_cwd.as_deref().unwrap_or("")
        } else {
            &cwd
        };

        let session_ids: Vec<String> = sessions.iter().map(|(id, _)| id.clone()).collect();
        let most_recent = sessions.iter().map(|(_, t)| *t).fold(0.0f64, f64::max);
        let created_at = sessions.iter().map(|(_, t)| *t).fold(f64::MAX, f64::min);

        let composite_id = registry.register(encoded_name, cwd_str, session_ids.clone());

        projects.push(Project {
            id: composite_id,
            path: cwd_str.to_string(),
            name: extract_project_name(encoded_name, Some(cwd_str)),
            sessions: session_ids,
            created_at: if created_at == f64::MAX { 0.0 } else { created_at },
            most_recent_session: if most_recent > 0.0 {
                Some(most_recent)
            } else {
                None
            },
        });
    }

    Ok(projects)
}

/// Extract the `cwd` field from the first line of a JSONL file.
fn extract_cwd_from_file(file_path: &Path) -> Option<String> {
    let file = std::fs::File::open(file_path).ok()?;
    let reader = BufReader::new(file);

    for line in reader.lines() {
        let line = line.ok()?;
        if line.trim().is_empty() {
            continue;
        }
        if let Ok(entry) = serde_json::from_str::<RawJsonlEntry>(&line) {
            if let Some(cwd) = entry.cwd {
                if !cwd.is_empty() {
                    return Some(cwd);
                }
            }
        }
        // Only check the first non-empty line
        break;
    }

    None
}

File: internal/discovery/project_scanner.go
package discovery

// Project scanning — discover projects from ~/.claude/projects/.
// Mirrors src-tauri/src/discovery/project_scanner.rs.

import (
	"bufio"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"claude-devtools/internal/domain"
)

// ScanProjects scans the projects directory and returns all projects sorted by
// most-recent-session descending. Mirrors project_scanner::scan_projects.
func ScanProjects(projectsDir string, registry *SubprojectRegistry) ([]domain.Project, error) {
	if _, err := os.Stat(projectsDir); os.IsNotExist(err) {
		return []domain.Project{}, nil
	}

	registry.Clear()
	var allProjects []domain.Project

	entries, err := os.ReadDir(projectsDir)
	if err != nil {
		return nil, fmt.Errorf("failed to read projects dir: %w", err)
	}

	for _, e := range entries {
		if !e.IsDir() || !IsValidEncodedPath(e.Name()) {
			continue
		}
		projects, err := scanProject(projectsDir, e.Name(), registry)
		if err != nil {
			// Log and continue, matching Rust's tracing::warn behaviour.
			continue
		}
		allProjects = append(allProjects, projects...)
	}

	// Sort by most_recent_session descending.
	sortProjectsByMostRecent(allProjects)
	return allProjects, nil
}

// scanProject scans a single project directory. May return multiple projects if
// sessions have different cwd values (subproject splitting).
func scanProject(projectsDir, encodedName string, registry *SubprojectRegistry) ([]domain.Project, error) {
	projectDir := filepath.Join(projectsDir, encodedName)
	entries, err := os.ReadDir(projectDir)
	if err != nil {
		return nil, fmt.Errorf("failed to read %s: %w", projectDir, err)
	}

	// sessions by cwd: cwd → [(sessionID, createdAt)]
	type sessionInfo struct {
		id        string
		createdAt float64
	}
	sessionsByCWD := map[string][]sessionInfo{}
	var defaultCWD string

	for _, e := range entries {
		name := e.Name()
		if e.IsDir() || !strings.HasSuffix(name, ".jsonl") || strings.HasPrefix(name, "agent_") {
			continue
		}

		sessionID := strings.TrimSuffix(name, ".jsonl")
		filePath := filepath.Join(projectDir, name)

		// mtime as created_at (matches session_lister which also uses mtime).
		info, err := e.Info()
		var createdAt float64
		if err == nil {
			createdAt = float64(info.ModTime().UnixMilli())
		}

		cwd := extractCWDFromFile(filePath)

		if defaultCWD == "" && cwd != "" {
			defaultCWD = cwd
		}

		key := cwd
		if key == "" {
			key = "__default__"
		}
		sessionsByCWD[key] = append(sessionsByCWD[key], sessionInfo{id: sessionID, createdAt: createdAt})
	}

	// Single cwd group (or empty) → one project.
	if len(sessionsByCWD) <= 1 {
		var sessions []sessionInfo
		var cwdKey string
		for k, v := range sessionsByCWD {
			cwdKey = k
			sessions = v
			break
		}

		var cwdHint string
		if cwdKey == "__default__" || cwdKey == "" {
			cwdHint = defaultCWD
		} else {
			cwdHint = cwdKey
		}

		sessionIDs := make([]string, 0, len(sessions))
		mostRecent := 0.0
		earliest := -1.0
		for _, s := range sessions {
			sessionIDs = append(sessionIDs, s.id)
			if s.createdAt > mostRecent {
				mostRecent = s.createdAt
			}
			if earliest < 0 || s.createdAt < earliest {
				earliest = s.createdAt
			}
		}
		if earliest < 0 {
			earliest = 0
		}

		path := cwdHint
		if path == "" {
			path = DecodePath(encodedName)
		}

		var mostRecentPtr *float64
		if mostRecent > 0 {
			mostRecentPtr = &mostRecent
		}

		return []domain.Project{{
			ID:                encodedName,
			Path:              path,
			Name:              ExtractProjectName(encodedName, cwdHint),
			Sessions:          sessionIDs,
			CreatedAt:         earliest,
			MostRecentSession: mostRecentPtr,
		}}, nil
	}

	// Multiple cwds → create composite projects.
	var projects []domain.Project
	for cwdKey, sessions := range sessionsByCWD {
		cwdStr := cwdKey
		if cwdStr == "__default__" {
			cwdStr = defaultCWD
		}

		sessionIDs := make([]string, 0, len(sessions))
		mostRecent := 0.0
		earliest := -1.0
		for _, s := range sessions {
			sessionIDs = append(sessionIDs, s.id)
			if s.createdAt > mostRecent {
				mostRecent = s.createdAt
			}
			if earliest < 0 || s.createdAt < earliest {
				earliest = s.createdAt
			}
		}
		if earliest < 0 {
			earliest = 0
		}

		compositeID := registry.Register(encodedName, cwdStr, sessionIDs)

		var mostRecentPtr *float64
		if mostRecent > 0 {
			mostRecentPtr = &mostRecent
		}

		projects = append(projects, domain.Project{
			ID:                compositeID,
			Path:              cwdStr,
			Name:              ExtractProjectName(encodedName, cwdStr),
			Sessions:          sessionIDs,
			CreatedAt:         earliest,
			MostRecentSession: mostRecentPtr,
		})
	}

	return projects, nil
}

// extractCWDFromFile reads the cwd field from the first non-empty JSONL line.
func extractCWDFromFile(filePath string) string {
	f, err := os.Open(filePath)
	if err != nil {
		return ""
	}
	defer f.Close()

	scanner := bufio.NewScanner(f)
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if line == "" {
			continue
		}
		var entry struct {
			Cwd *string `json:"cwd"`
		}
		if err := json.Unmarshal([]byte(line), &entry); err == nil {
			if entry.Cwd != nil && *entry.Cwd != "" {
				return *entry.Cwd
			}
		}
		// Only check the first non-empty line.
		break
	}
	return ""
}

// sortProjectsByMostRecent sorts in-place by MostRecentSession descending.
func sortProjectsByMostRecent(projects []domain.Project) {
	for i := 1; i < len(projects); i++ {
		for j := i; j > 0; j-- {
			aTime := 0.0
			if projects[j].MostRecentSession != nil {
				aTime = *projects[j].MostRecentSession
			}
			bTime := 0.0
			if projects[j-1].MostRecentSession != nil {
				bTime = *projects[j-1].MostRecentSession
			}
			if aTime > bTime {
				projects[j], projects[j-1] = projects[j-1], projects[j]
			} else {
				break
			}
		}
	}
}
